README.md
# Xelera Decision Tree Inference

**Xelera Decision Tree Inference** provides FPGA-accelerated inference (prediction) for real-time Classification and Regression applications when high-throughput or low-latency matters. It supports **Random Forest**, **XGBoost** and **LightGBM** algorithms.

1. Train your own model using one of the supported frameworks (**scikit-learn**, **XGBoost**, **LightGBM**, **H20.ai**) and convert it to a unified representation (XlModel) for Alveo Accelerator Cards

<p align="center">
<img src="docs/images/flow0.png" align="middle" width="500"/>
</p>

2. Integrate with your application via Python and run with an auto-scalable inference server

<p align="center">
<img src="docs/images/flow1.png" align="middle" width="500"/>
</p>


Additional resource:
* [Random Forest Inference Benchmark on Lenovo Thinksystem SE350](https://xelera.io/assets/downloads/Benchmarks/benchmark-001---edge-server-random-forest-inference.pdf)
* [Blog post on accelerating Decision tree-based predictive analytics](https://xelera.io/blog/acceleration-of-Decision-tree-ensembles)

## What's New on 0.6.0b6drm
[Release notes](docs/releaseNotes.md)
* DRM-based licensing system
* No feature scaling required: float32-based tree traversal algorithm in FPGA
* Kernel optimized for large ensambles and RF classification (greater than hundreds of trees)

## Acceleration Platforms

|            Cards/Platform            |     Shell        |
| :-------------------------: |:-------------------------: |
|   [Xilinx Alveo U50](https://www.xilinx.com/products/boards-and-kits/alveo/u50.html) | xilinx_u50_gen3x16_xdma_201920_3 |  
|   [Xilinx Alveo U200](https://www.xilinx.com/products/boards-and-kits/alveo/u200.html) | xilinx-u200-xdma-201830.2 | 
|   [AWS f1.2xlarge](https://aws.amazon.com/de/ec2/instance-types/f1/)                     | xilinx_aws-vu9p-f1_shell-v04261818_201920_1 |

## Features and Limitations
For supported features and current limitations, see [supported parameters](docs/supportedFeatures.md).

## Usage

#### Installation

Xelera Decision Tree Inference is available:
* [On-premises](docs/on-premises.md)
* [AWS](docs/aws-marketplace.md)

#### Get started with examples
* [Predict the flight delay](docs/exampleFlight.md)

## Cheat Sheet

* [Random Forest](docs/cheatSheetRF.md)
* [XGBoost](docs/cheatSheetXGBoost.md)
* [LightGBM](docs/cheatSheetLightGBM.md)

## API changes

See [API migration](docs/migration.md) for instructions to migrate to the updated API of this release.

## Contacts

In case of questions, contact [info@xelera.io](mailto:info@xelera.io)


docs/releaseNotes.md
### 0.6.0b6drm (Beta)
* Supported algorithms:
    * Random Forest
        * Regression
        * Classification (Binomial and Multinomial)
    * XGBoost
        * Regression
        * Classification (Binomial and Multinomial on selected Alveo cards)
    * LightGBM
        * Regression
        * Classification (Binomial and Multinomial on selected Alveo cards)
* Supported frameworks:
    * scikit-learn
    * XGBboost
    * LightGBM
* No feature scaling required: float32-based tree traversal algorithm in FPGA
* Kernel optimized for large ensambles and RF classification (greater than hundreds of trees) 
* Single model inference
* Latency optimized inference
* Python interface with non-blocking  calls
* Inference server with a single FPGA card
* DRM-based licensing system

### 0.4.0b4 (Beta)
* Supported algorithms:
    * Random Forest
        * Regression
        * Classification (Binomial and Multinomial)
    * XGBoost
        * Regression
        * Classification (Binomial and Multinomial on selected Alveo cards)
    * LightGBM
        * Regression
        * Classification (Binomial and Multinomial on selected Alveo cards)
* Supported frameworks:
    * scikit-learn
    * XGBboost
    * LightGBM
* Single model inference
* Latency optimized inference
* Python interface with non-blocking  calls
* Inference server with a single FPGA card
* License support

### 0.3.0b3 (Beta)
* Supported algorithms:
    * Random Forest
        * Regression
        * Classification (Binomial and Multinomial)
    * XGBoost
        * Regression
        * Classification (Binomial)
    * LightGBM
        * Regression
        * Classification (Binomial)
* Supported frameworks:
    * scikit-learn
    * XGBboost
    * LightGBM
* Python interface
* Single model inference
* Latency optimized inference (no throughput mode)


scripts/utils/arguments.py
#!/usr/bin/python3

import argparse

def str2bool(value):
    """A function to convert string to bool value."""
    if value.lower() in {'yes', 'true', 't', 'y', '1'}:
        return True
    if value.lower() in {'no', 'false', 'f', 'n', '0'}:
        return False
    raise argparse.ArgumentTypeError('Boolean value expected.')

def parse_arguments():
    """Parse arguments.

    Returns:
    A tuple of:
        - `tree_inference_args`: all arguments for tree-based inferencing
    """

    # tree_inference configurations
    tree_inference_parser = argparse.ArgumentParser(
        description='Common configurations.', add_help=False)

    tree_inference_parser.add_argument(
        '--enable_regression',
        default='true',
        type=str2bool,
        help='Enables regression with numerical features. Default: true')
    tree_inference_parser.add_argument(
        '--enable_binomial',
        default='false',
        type=str2bool,
        help='Enables binomial with numerical features. Default: false')
    tree_inference_parser.add_argument(
        '--enable_multinomial',
        default='false',
        type=str2bool,
        help='Enables multinomial with numerical features. Default: false')
    tree_inference_parser.add_argument(
        '--enable_SW_inference',
        default='true',
        type=str2bool,
        help='Enables CPU-based inference. Default: true')
    tree_inference_parser.add_argument(
        '--enable_FPGA_inference',
        default='true',
        type=str2bool,
        help='Enables FPGA-based inference. Default: true')
    tree_inference_parser.add_argument(
        '--max_depth',
        type=int,
        default=8,
        help='Maximum number of tree levels for training. Default: 8')
    tree_inference_parser.add_argument(
        '--number_of_trees',
        type=int,
        default=100,
        help='Number of trees per forest for training. Default: 100')
    tree_inference_parser.add_argument(
        '--num_test_samples',
        type=int,
        default=100,
        help='Number of samples per forest for inference. Default: 100')
    tree_inference_parser.add_argument(
        '--n_loops',
        type=int,
        default=1000,
        help='Number of iterations over the same data for a better average execution time. Default: 1000')
    tree_inference_parser.add_argument(
        '--data_fpath',
        type=str,
        default='/app/xelera_demo/data/flight-delays/flights.csv',
        help='Dataset file. Default: ./data/flight-delays/flights.csv')


    # a super parser for sanity checks
    super_parser = argparse.ArgumentParser(
        parents=[tree_inference_parser])

    # get arguments
    super_parser.parse_args()
    tree_inference_args, _ = tree_inference_parser.parse_known_args()

    return (tree_inference_args)


scripts/demo_CPU_XGB_regression_flight.py
import pandas as pd, numpy as np, time
from sklearn.model_selection import train_test_split
from sklearn.metrics import auc, accuracy_score, confusion_matrix, mean_squared_error, roc_auc_score
import xgboost as xgb
import time
import pickle
import os
import sys
from xgboost import DMatrix as DMatrix
from utils import arguments

tree_inference_args = arguments.parse_arguments()


######Free parameters######
max_depth = tree_inference_args.max_depth
number_of_trees = tree_inference_args.number_of_trees
numTestSamples = tree_inference_args.num_test_samples
nLoops = tree_inference_args.n_loops
dataset_name = tree_inference_args.data_fpath

modelFileName = "XGB_Regression_Flight_" + str(number_of_trees) + "trees.pkl"
##############################

print("Loading dataset ...")

data_origin = pd.read_csv(dataset_name)
data_origin = data_origin.sample(frac = 0.1, random_state=10)

feature_names = ["MONTH","DAY","DAY_OF_WEEK","AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT",
                 "ORIGIN_AIRPORT","AIR_TIME", "DEPARTURE_TIME","DISTANCE","ARRIVAL_DELAY"]

data_origin = data_origin[feature_names]
data_origin.dropna(inplace=True)

feature_names.remove("ARRIVAL_DELAY") # its the target

num_classes = 1
max_depth = 8

data = data_origin.copy()

cols = ["AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT","ORIGIN_AIRPORT"]
for item in cols:
    data[item] = data[item].astype("category").cat.codes +1


data_label = data["ARRIVAL_DELAY"]
data = data.drop(["ARRIVAL_DELAY"], axis=1)

print(data.head(5) )

x_train, x_test, y_train, y_test = train_test_split(data, data_label,random_state=10, test_size=0.25)

if numTestSamples <= y_test.shape[0]:
    x_test = x_test.head(numTestSamples)
    y_test = y_test.head(numTestSamples)
else:
    numTestSamples = y_test.shape[0]

######### Training ########

if (os.path.isfile(modelFileName)):
    print("Model is already available. Training is NOT run")
else:
    print("Model is not available, running training ...")

    model = xgb.XGBRegressor(objective="reg:squarederror", booster="gbtree", random_state=42, max_depth=max_depth, n_estimators=number_of_trees, base_score=0.0, n_jobs = 32)
    model.fit(x_train,y_train)

    # save the trained xgboost model
    pickle.dump(model, open(modelFileName, 'wb'))

    del model


######### Inference ########
print("Running Inference ...")

try:
    f = open(modelFileName, 'rb')
    model = pickle.load(f)
except IOError:
    print("File",modelFileName," not accessible")
finally:
    f.close()

start_time = time.perf_counter()
for i in range(nLoops):
    # To avoid caching and to have more accurate performance numbers, pass a new array every time
    y_pred = model.predict(x_test + i)
stop_time = time.perf_counter()
execution_time = stop_time-start_time

mse=mean_squared_error(y_test, y_pred)

del model

print("-----------------------")
print("INFO: Number of features:",x_test.shape[1])
print("INFO: Number of samples:", y_pred.size)
print("INFO: Number of trees:",number_of_trees)

print("INFO: CPU mse:",np.sqrt(mse))

print("INFO: CPU throughput:", "{:.2e}".format(nLoops*y_pred.size/execution_time), "samples/s")

scripts/demo_FPGA_XGB_regression_flight.py
import pandas as pd, numpy as np, time
from sklearn.model_selection import train_test_split
from sklearn.metrics import auc, accuracy_score, confusion_matrix, mean_squared_error, roc_auc_score
import xgboost as xgb
import time
import pickle
import os
import sys
from xgboost import DMatrix as DMatrix
from utils import arguments

import XlPluginDecisionTreeInference as xl

tree_inference_args = arguments.parse_arguments()


######Free parameters######
max_depth = tree_inference_args.max_depth
number_of_trees = tree_inference_args.number_of_trees
numTestSamples = tree_inference_args.num_test_samples
nLoops = tree_inference_args.n_loops
dataset_name = tree_inference_args.data_fpath

modelFileName = "./XGB_Regression_Flight_" + str(number_of_trees) + "trees.pkl"
modelFPGAFileName = "./XGB_Regression_Flight_" + str(number_of_trees) + "trees.xlmodel"
##############################

print("Loading dataset ...")

data_origin = pd.read_csv(dataset_name)
data_origin = data_origin.sample(frac = 0.1, random_state=10)

feature_names = ["MONTH","DAY","DAY_OF_WEEK","AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT",
                 "ORIGIN_AIRPORT","AIR_TIME", "DEPARTURE_TIME","DISTANCE","ARRIVAL_DELAY"]

data_origin = data_origin[feature_names]
data_origin.dropna(inplace=True)

feature_names.remove("ARRIVAL_DELAY") # its the target

num_classes = 1
max_depth = 8

data = data_origin.copy()

cols = ["AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT","ORIGIN_AIRPORT"]
for item in cols:
    data[item] = data[item].astype("category").cat.codes +1


data_label = data["ARRIVAL_DELAY"]
data = data.drop(["ARRIVAL_DELAY"], axis=1)

print(data.head(5) )

x_train, x_test, y_train, y_test = train_test_split(data, data_label,random_state=10, test_size=0.25)

if numTestSamples <= y_test.shape[0]:
    x_test = x_test.head(numTestSamples)
    y_test = y_test.head(numTestSamples)
else:
    numTestSamples = y_test.shape[0]

######### Training ########

if (os.path.isfile(modelFileName)):
    print("Model is already available. Training is NOT run")
else:
    print("Model is not available, start training ...")

    model = xgb.XGBRegressor(objective="reg:squarederror", booster="gbtree", random_state=42, max_depth=max_depth, n_estimators=number_of_trees, base_score=0.0, n_jobs = 32)
    model.fit(x_train,y_train)

    # save the trained xgboost model
    pickle.dump(model, open(modelFileName, 'wb'))

    del model


if (os.path.isfile(modelFPGAFileName)):
    print ("FPGA Model is already available. xlmodel conversion is NOT run")
else:
    print("FPGA Model is not available, running xlmodel conversion ...")

    try:
        f = open(modelFileName, 'rb')
        model = pickle.load(f)
    except IOError:
        print("File",modelFileName," not accessible")
    finally:
        f.close()

    #convert model for FPGA
    xlSetup = xl.XlXGBSetup()
    fpga_model = xlSetup.getModelForFPGA(model.get_booster(), max_depth, num_classes, feature_names, modelFPGAFileName)

    del xlSetup
    del fpga_model
    del model

######### Inference ########
print("Running Inference ...")

try:  
    inference_engine = xl.XlRFInference()
    inference_engine.setModel(modelFPGAFileName)
except:
    print("File",modelFPGAFileName," not accessible")

start_time = time.perf_counter()
x_test_np = np.array(x_test, dtype=np.float32, order='C')
for i in range(nLoops):
    # To avoid caching and to have more accurate performance numbers, pass a new array every time 
    inference_engine.predict(x_test_np + i)
for i in range(nLoops):
    y_pred = inference_engine.get_results()
stop_time = time.perf_counter()
execution_time = stop_time-start_time

mse=mean_squared_error(y_test, y_pred)
del inference_engine
xl.shutdown()


##############################
print("INFO: Number of features:",x_test.shape[1])
print("INFO: Number of samples:", y_pred.size)
print("INFO: Number of trees:",number_of_trees)

print("INFO: FPGA mse:",np.sqrt(mse))

print("INFO: FPGA throughput:", "{:.2e}".format(nLoops*y_pred.size/execution_time), "samples/s")




scripts/RF_scikit_flight.py
import pandas as pd, numpy as np, time
from sklearn.model_selection import train_test_split
from sklearn.metrics import auc, accuracy_score, confusion_matrix, mean_squared_error, roc_auc_score
from sklearn.ensemble import RandomForestClassifier as skrfc
from sklearn.ensemble import RandomForestRegressor as skrfr
import pickle
import os
import time
import sys


from utils import arguments

tree_inference_args = arguments.parse_arguments()

enable_regression = tree_inference_args.enable_regression
enable_binomial = tree_inference_args.enable_binomial
enable_multinomial = tree_inference_args.enable_multinomial

enable_SW_inference = tree_inference_args.enable_SW_inference
enable_FPGA_inference = tree_inference_args.enable_FPGA_inference


max_depth = tree_inference_args.max_depth
number_of_trees = tree_inference_args.number_of_trees
numTestSamples = tree_inference_args.num_test_samples

nLoops = tree_inference_args.n_loops

print("Loading dataset...")

dataset_name = tree_inference_args.data_fpath
data_origin = pd.read_csv(dataset_name)
if (enable_FPGA_inference):
    import XlPluginDecisionTreeInference as xl

data_origin = data_origin.sample(frac = 0.1, random_state=10)

data_origin = data_origin[["MONTH","DAY","DAY_OF_WEEK","AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT",
                 "ORIGIN_AIRPORT","AIR_TIME", "DEPARTURE_TIME","DISTANCE","ARRIVAL_DELAY"]]
data_origin.dropna(inplace=True)


data_origin = data_origin.head(300000)





############################
## Regression
############################

if enable_regression:

    print("################################################")
    print("RF Regression with Numerical Features")

    modelFileName = "./RF_Regression_Flight_" + str(number_of_trees) + "trees.pkl"
    modelFPGAFileName = "./RF_Regression_Flight_" + str(number_of_trees) + "trees.xlmodel"

    data = data_origin.copy()


    cols = ["AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT","ORIGIN_AIRPORT"]
    for item in cols:
        data[item] = data[item].astype("category").cat.codes +1


    data_label = data["ARRIVAL_DELAY"]
    data = data.drop(["ARRIVAL_DELAY"], axis=1)

    data = pd.DataFrame(data.astype(np.float32))
    # cuML Random Forest Classifier requires the labels to be integers
    data_label = pd.Series(data_label.astype(np.int32))


    x_train, x_test, y_train, y_test = train_test_split(data, data_label,random_state=10, test_size=0.25)

    if numTestSamples <= y_test.shape[0]:
        x_test = x_test.head(numTestSamples)
        y_test = y_test.head(numTestSamples)
    else:
        numTestSamples = y_test.shape[0]
        print("numTestSamples capped to", numTestSamples)


    ######### Training ########

    #set "n_estimators": 300 to get good results
    model_sk_regression = skrfr(max_depth=max_depth,n_estimators=number_of_trees, n_jobs=32, random_state=1234)
    
    if (os.path.isfile(modelFileName)):
        print("Model is already available. Training is NOT run")
    else:
        print("Model is not available, start training ...")
    
        start_time = time.perf_counter()
        model_sk_regression.fit(x_train, y_train)

        print("Training_time:      ", time.perf_counter() - start_time, "s")
        pickle.dump(model_sk_regression, open(modelFileName, 'wb'))
        

        import XlPluginDecisionTreeInference as xl
        start_time = time.perf_counter()
        xlrfsetup = xl.XlRandomForestSetup()
        xlrfsetup.getModelForFPGA(model_sk_regression,modelFPGAFileName)
        del xlrfsetup
        print("model conversion to .xlmodel time:", time.perf_counter() - start_time, "s")



    ######### Inference SW ########

    if enable_SW_inference:
        
        sw_time = 0
        print("Starting SW inference...")

        try:
            f = open(modelFileName, 'rb')
            model_sk_regression = pickle.load(f)
        except IOError:
            print("File",modelFileName," not accessible")
        finally:
            f.close()

        #throughput and latency measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            y_pred = model_sk_regression.predict(x_test+i)
        stop_time = time.perf_counter()
        sw_time = stop_time - start_time

        mse=mean_squared_error(y_test, y_pred)

        print("SW mse",np.sqrt(mse))
        print(y_test)
        print(y_pred)
        error = abs(y_test - y_pred)
        print("SW error",error.mean())
        print('SW predict latency (average on', nLoops,'runs for', y_pred.size, 'samples): ', "{:.2e}".format(sw_time/nLoops), 's')
        print("SW predict throughput:", "{:.2e}".format(nLoops*y_pred.size/sw_time), "samples/s")
        print("SW Number of features:",x_test.shape[1])
        print("SW Number of trees:",number_of_trees)


    ######### Inference FPGA ########

    if (enable_FPGA_inference):
        import XlPluginDecisionTreeInference as xl

        setup_starttime = time.perf_counter()

        xlrf = xl.XlRFInference()
        xlrf.setModel(modelFPGAFileName)

        setup_endtime = time.perf_counter()
        HW_setup_time = setup_endtime - setup_starttime

        x_nd = np.array(x_test, dtype=np.float32, order='C')

        print("Starting HW inference ...")
        #throughput measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            xlrf.predict(x_nd+i)
        for i in range(nLoops):
            y_pred = xlrf.get_results()
        stop_time = time.perf_counter()
        hw_time_throughput = stop_time - start_time

        #latency measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            xlrf.predict(x_nd+i)
            y_pred = xlrf.get_results()
        stop_time = time.perf_counter()
        hw_time_latency = stop_time - start_time

        mse=mean_squared_error(y_test, y_pred)
        print("HW mse",np.sqrt(mse))
        error = abs(y_test - y_pred)
        print("HW error",error.mean())
        print('HW predict latency (average on', nLoops,'runs for', y_pred.size, 'samples): ', "{:.2e}".format(hw_time_latency/nLoops), 's')
        print("HW predict throughput:", "{:.2e}".format(nLoops*y_pred.size/hw_time_throughput), "samples/s")
        print("HW Number of features:",x_test.shape[1])
        print("HW Number of trees:",number_of_trees)
        del xlrf

############################
## Binomial
############################

if enable_binomial:

    print("##############################################")
    print("RF Binomial with Numerical Features")

    modelFileName = "./RF_Binomial_Flight_" + str(number_of_trees) + "trees.pkl"
    modelFPGAFileName = "./RF_Binomial_Flight_" + str(number_of_trees) + "trees.xlmodel"

    data = data_origin.copy()

    cols = ["AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT","ORIGIN_AIRPORT"]
    for item in cols:
        data[item] = data[item].astype("category").cat.codes +1

    #create 2 classes for the label
    data['ARRIVAL_DELAY'] = np.where(data['ARRIVAL_DELAY'] <= 10, 0, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where(data['ARRIVAL_DELAY'] > 10, 1, data['ARRIVAL_DELAY'])

    data_label = data["ARRIVAL_DELAY"]
    data = data.drop(["ARRIVAL_DELAY"], axis=1)

    data = pd.DataFrame(data.astype(np.float32))
    # cuML Random Forest Classifier requires the labels to be integers
    data_label = pd.Series(data_label.astype(np.int32))

    x_train, x_test, y_train, y_test = train_test_split(data, data_label,random_state=10, test_size=0.25)

    if numTestSamples <= y_test.shape[0]:
        x_test = x_test.head(numTestSamples)
        y_test = y_test.head(numTestSamples)
    else:
        numTestSamples = y_test.shape[0]
        print("numTestSamples capped to", numTestSamples)


    ######### Training ########

    print("SK binomial: Start training ...")

    #set iterations= 500 to get good results
    model_sk_binomial = skrfc(max_depth=max_depth,n_estimators=number_of_trees, n_jobs=32, random_state=1234)

    if (os.path.isfile(modelFileName)):
        print("Model is already available. Training is NOT run")
    else:
        print("Model is not available, start training ...")

        start_time = time.perf_counter()
        model_sk_binomial.fit(x_train,y_train)
        print("Training_time:      ", time.perf_counter() - start_time, "s")
        pickle.dump(model_sk_binomial, open(modelFileName, 'wb'))

        import XlPluginDecisionTreeInference as xl

        start_time = time.perf_counter()
        xlrfsetup = xl.XlRandomForestSetup()
        xlrfsetup.getModelForFPGA(model_sk_binomial,modelFPGAFileName)
        del xlrfsetup
        print("model conversion to .xlmodel time:", time.perf_counter() - start_time, "s")


    ######### Inference SW ########

    if enable_SW_inference:

        print("Starting SW inference ...")

        try:
            f = open(modelFileName, 'rb')
            model_sk_binomial = pickle.load(f)
        except IOError:
            print("File",modelFileName," not accessible")
        finally:
            f.close()

        #throughput and latency measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            y_pred = model_sk_binomial.predict(x_test+i)
        stop_time = time.perf_counter()
        sw_time = stop_time - start_time

        mse=mean_squared_error(y_test, y_pred)

        print("SW mse",np.sqrt(mse))
        error = abs(y_test - y_pred)
        print("SW error",error.mean())
        print("SW accuracy score",accuracy_score(y_test, y_pred))
        print('SW predict latency (average on', nLoops,'runs for', y_pred.size, 'samples): ', "{:.2e}".format(sw_time/nLoops), 's')
        print("SW predict throughput:", "{:.2e}".format(nLoops*y_pred.size/sw_time), "samples/s")
        print("SW Number of features:",x_test.shape[1])
        print("SW Number of trees:",number_of_trees)
        print("SW Number of classes:",model_sk_binomial.classes_.shape[0])


    ######### Inference FPGA ########

    if (enable_FPGA_inference):
        import XlPluginDecisionTreeInference as xl

        setup_starttime = time.perf_counter()

        xlrf = xl.XlRFInference()
        xlrf.setModel(modelFPGAFileName)

        setup_endtime = time.perf_counter()
        HW_setup_time = setup_endtime - setup_starttime

        x_nd = np.array(x_test, dtype=np.float32, order='C')
        print("Starting HW inference ...")
        #throughput measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            xlrf.predict(x_nd+i)
        for i in range(nLoops):
            y_pred = xlrf.get_results()
        stop_time = time.perf_counter()
        hw_time_throughput = stop_time - start_time

        #latency measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            xlrf.predict(x_nd+i)
            y_pred = xlrf.get_results()
        stop_time = time.perf_counter()
        hw_time_latency = stop_time - start_time

        y_pred = y_pred.argmax(1)
        y_pred = model_sk_binomial.classes_.take(y_pred, axis=0)

        y_test = y_test.astype(int)


        mse=mean_squared_error(y_test, y_pred)
        print("HW mse",np.sqrt(mse))
        error = abs(y_test - y_pred)
        print("HW error",error.mean())
        print("HW accuracy score",accuracy_score(y_test, y_pred))
        print('HW predict latency (average on', nLoops,'runs for', y_pred.size, 'samples): ', "{:.2e}".format(hw_time_latency/nLoops), 's')
        print("HW predict throughput:", "{:.2e}".format(nLoops*y_pred.size/hw_time_throughput), "samples/s")
        print("HW Number of features:",x_test.shape[1])
        print("HW Number of trees:",number_of_trees)
        print("HW Number of classes:",model_sk_binomial.classes_.shape[0])
        
        del xlrf


############################
## Multinomial
############################

if enable_multinomial:

    print("##############################################")
    print("RF Multinomial with Numerical Features")

    modelFileName = "./RF_Multinomial_Flight_" + str(number_of_trees) + "trees.pkl"
    modelFPGAFileName = "./RF_Multinomial_Flight_" + str(number_of_trees) + "trees.xlmodel"

    data = data_origin.copy()

    cols = ["AIRLINE","FLIGHT_NUMBER","DESTINATION_AIRPORT","ORIGIN_AIRPORT"]
    for item in cols:
        data[item] = data[item].astype("category").cat.codes

    #create 10 classes for the label
    data['ARRIVAL_DELAY'] = np.where(data['ARRIVAL_DELAY'] <= 0, 0, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>0 ) & (data['ARRIVAL_DELAY']<=1), 1, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>1 ) & (data['ARRIVAL_DELAY']<=2), 2, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>2 ) & (data['ARRIVAL_DELAY']<=3), 3, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>3 ) & (data['ARRIVAL_DELAY']<=4), 4, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>4 ) & (data['ARRIVAL_DELAY']<=5), 5, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>5 ) & (data['ARRIVAL_DELAY']<=6), 6, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>6 ) & (data['ARRIVAL_DELAY']<=7), 7, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where((data['ARRIVAL_DELAY']>7 ) & (data['ARRIVAL_DELAY']<=8), 8, data['ARRIVAL_DELAY'])
    data['ARRIVAL_DELAY'] = np.where(data['ARRIVAL_DELAY'] > 8, 9, data['ARRIVAL_DELAY'])

    data_label = data["ARRIVAL_DELAY"] + 0
    # data_label = data["ARRIVAL_DELAY"].astype("category").cat.codes +1
    data = data.drop(["ARRIVAL_DELAY"], axis=1)

    data = pd.DataFrame(data.astype(np.float32))
    data_label = pd.Series(data_label.astype(np.int32))

    x_train, x_test, y_train, y_test = train_test_split(data, data_label,random_state=10, test_size=0.25)

    if numTestSamples <= y_test.shape[0]:
        x_test = x_test.head(numTestSamples)
        y_test = y_test.head(numTestSamples)
    else:
        numTestSamples = y_test.shape[0]
        print("numTestSamples capped to", numTestSamples)

    ######### Training ########

    print("SK multinomial: Start training ...")

    #set iterations= 500 to get good results
    model_sk_multinomial = skrfc(max_depth=max_depth,n_estimators=number_of_trees, n_jobs=32, random_state=1234)

    if (os.path.isfile(modelFileName)):
        print("Model is already available. Training is NOT run")
    else:
        print("Model is not available, start training ...")

        start_time = time.perf_counter()
        model_sk_multinomial.fit(x_train,y_train)
        print("Training_time:      ", time.perf_counter() - start_time, "s")
        pickle.dump(model_sk_multinomial, open(modelFileName, 'wb'))

        # if enable_FPGA_inference:
        import XlPluginDecisionTreeInference as xl


        start_time = time.perf_counter()
        xlrfsetup = xl.XlRandomForestSetup()
        xlrfsetup.getModelForFPGA(model_sk_multinomial,modelFPGAFileName)
        del xlrfsetup
        print("model conversion to .xlmodel time:", time.perf_counter() - start_time, "s")


    ######### Inference SW ########

    if enable_SW_inference:

        print("Starting SW inference ...")

        try:
            f = open(modelFileName, 'rb')
            model_sk_multinomial = pickle.load(f)
        except IOError:
            print("File",modelFileName," not accessible")
        finally:
            f.close()

        #throughput and latency measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            y_pred = model_sk_multinomial.predict(x_test+i)
        stop_time = time.perf_counter()
        sw_time = stop_time - start_time

        mse=mean_squared_error(y_test, y_pred)

        print("SW mse",np.sqrt(mse))
        error = abs(y_test - y_pred)
        print("SW error",error.mean())
        print("SW accuracy score",accuracy_score(y_test, y_pred))
        print('SW predict latency (average on', nLoops,'runs for', y_pred.size, 'samples): ', "{:.2e}".format(sw_time/nLoops), 's')
        print("SW predict throughput:", "{:.2e}".format(nLoops*y_pred.size/sw_time), "samples/s")
        print("SW Number of features:",x_test.shape[1])
        print("SW Number of trees:",number_of_trees)
        print("SW Number of classes:",model_sk_multinomial.classes_.shape[0])
    
    ######### Inference FPGA ########

    if (enable_FPGA_inference):
        import XlPluginDecisionTreeInference as xl

        setup_starttime = time.perf_counter()

        xlrf = xl.XlRFInference()
        xlrf.setModel(modelFPGAFileName)

        setup_endtime = time.perf_counter()
        HW_setup_time = setup_endtime - setup_starttime

        x_nd = np.array(x_test, dtype=np.float32, order='C')

        print("Starting HW inference ...")
        #throughput measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            xlrf.predict(x_nd+i)
        for i in range(nLoops):
            y_pred = xlrf.get_results()
        stop_time = time.perf_counter()
        hw_time_throughput = stop_time - start_time

        #latency measure
        start_time = time.perf_counter()
        for i in range(nLoops):
            xlrf.predict(x_nd+i)
            y_pred = xlrf.get_results()
        stop_time = time.perf_counter()
        hw_time_latency = stop_time - start_time

        y_pred = y_pred.argmax(1)
        y_pred = model_sk_multinomial.classes_.take(y_pred, axis=0)

        y_test = np.array(y_test)
        y_test = y_test.astype(int)


        mse=mean_squared_error(y_test, y_pred)
        print("HW mse",np.sqrt(mse))
        error = abs(y_test - y_pred)
        print("HW error",error.mean())
        print("HW accuracy score",accuracy_score(y_test, y_pred))
        print('HW predict latency (average on', nLoops,'runs for', y_pred.size, 'samples): ', "{:.2e}".format(hw_time_latency/nLoops), 's')
        print("HW predict throughput:", "{:.2e}".format(nLoops*y_pred.size/hw_time_throughput), "samples/s")
        print("HW Number of features:",x_test.shape[1])
        print("HW Number of trees:",number_of_trees)
        print("HW Number of classes:",model_sk_multinomial.classes_.shape[0])
        del xlrf

if (enable_FPGA_inference):
    xl.shutdown()
